src/Microsoft.Extensions.ClosedGenericMatcher.Sources/ClosedGenericMatcher.cs

```diff
@@ -40,16 +40,63 @@ public static Type ExtractGenericInterface(Type queryType, Type interfaceType)
                 throw new ArgumentNullException(nameof(interfaceType));
             }
 
-            Func<Type, bool> matchesInterface =
-                type => type.GetTypeInfo().IsGenericType && type.GetGenericTypeDefinition() == interfaceType;
-            if (matchesInterface(queryType))
+            // Checked type matches (i.e. is a closed generic type created from) the open generic type.
+            if (IsGenericInstantiation(queryType, interfaceType))
             {
-                // Checked type matches (i.e. is a closed generic type created from) the open generic type.
+
                 return queryType;
             }
 
             // Otherwise check all interfaces the type implements for a match.
-            return queryType.GetTypeInfo().ImplementedInterfaces.FirstOrDefault(matchesInterface);
+            // 
+            // We do this by recursing down to the most derived type, and then looking at the implemented
+            // interfaces on the way back out.
+            // - If multiple different generic instantiations exists, we want the most derived one.
+            // - If that doesn't break the tie, then we sort alphabetically so that it's deterministic.
+            return GetGenericInstantiation(queryType, interfaceType);
+        }
+
+        private static bool IsGenericInstantiation(Type candidate, Type interfaceType)
+        {
+            return
+                candidate.GetTypeInfo().IsGenericType &&
+                candidate.GetGenericTypeDefinition() == interfaceType;
+        }
+
+        private static Type GetGenericInstantiation(Type queryType, Type interfaceType)
+        {
+            // BaseType will be null for object and interfaces, which means we've reached 'bottom'.
+            var baseType = queryType?.GetTypeInfo().BaseType;
+
+            Type baseTypeMatch = null;
+            if (baseType != null)
+            {
+                baseTypeMatch = GetGenericInstantiation(baseType, interfaceType);
+            }
```

Comment:
Must be missing something. If we can get a non-`null` `bestMatch` at this level, shouldn't need to recurse. So why check `baseType` first?